core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -362,30 +372,105 @@ public long count() {
+               if (realIterator.hasNext()) {
+                  R value = realIterator.next();
+                  segmentTracker.returningObject(value);
+                  return value;
+               } else {
+                  segmentTracker.onComplete();
+               }
+               return null;
+            }
+         };
+      }
+      return realIterator;
+   }
+
+   /**
+    * Tracking class that keeps track of segment completions and maps them to a given value. This value is not actually
+    * part of these segments, but is instead the object returned immediately after the segments complete. This way
+    * we can guarantee to notify the user after all elements have been processed of which segments were completed.
+    * All methods except for accept(int) are guaranteed to be called sequentially and in a safe manner.
+    */
+   private class CompletionSegmentTracker implements IntConsumer, io.reactivex.functions.Consumer<Object> {
+      private final Consumer<Supplier<PrimitiveIterator.OfInt>> listener;
+      private final Map<Object, IntSet> awaitingNotification;
+      volatile IntSet completedSegments;
+
+      private CompletionSegmentTracker(Consumer<Supplier<PrimitiveIterator.OfInt>> listener) {
+         this.listener = Objects.requireNonNull(listener);
+         this.awaitingNotification = new HashMap<>();
+         this.completedSegments = IntSets.concurrentSet(maxSegment);
+      }
+
+
+      @Override
+      public void accept(int value) {
+         // This method can technically be called from multiple threads
+         completedSegments.set(value);
+      }
+
+      @Override
+      public void accept(Object r) {
+         if (!completedSegments.isEmpty()) {
+            log.tracef("Going to complete segments %s when %s is iterated upon", completedSegments, Util.toStr(r));
+            awaitingNotification.put(r, completedSegments);
+            completedSegments = IntSets.concurrentSet(maxSegment);
          }
-         onClose(closeableIterator::close);
-         // This cast messes up generic checking, but that is okay
-         Function<R, R> function = iteratorOperation.getFunction();
-         if (function != null) {
-            return new IteratorMapper<>(closeableIterator, function);
-         } else {
-            return closeableIterator;
+      }
+
+      public void returningObject(Object value) {
+         IntSet segments = awaitingNotification.remove(value);
+         if (segments != null) {
+            log.tracef("Notifying listeners of segments %s complete now that %s is returning", segments, Util.toStr(value));
+            listener.accept(segments::iterator);
          }
       }
+
+      public void onComplete() {
+         log.tracef("Completing last segments of: %s", completedSegments);
+         listener.accept(completedSegments::iterator);
+         completedSegments.clear();
+      }
    }
 
    private class RehashIterator<S> extends AbstractIterator<S> implements CloseableIterator<S> {
```

Comment:
It seems we can get rid of this and `CompletionListenerRehashIterator`

Reply:
Yup that will be handled in https://issues.jboss.org/browse/ISPN-10895. I didn't want to bloat this PR too much.